In [6]:
def make_integer(nums, length):
    nums.sort()
    return "".join([str(n) for n in nums[:length]])

In [7]:
nums = [3, 1, 2, 6, 5, 9] 
length = 3
make_integer(nums, length)

'123'

In [30]:
def fix_misspellings(corrections):
    return [d['word'][:d['position']-1]+d['word'][d['position']:] for d in corrections]

In [31]:
corrections = [ { "word": "tablett", "position": 7 },
  { "word": "marrble", "position": 4 },
  { "word": "xdocker", "position": 1 },
]

In [32]:
fix_misspellings(corrections)

['tablet', 'marble', 'docker']

There is no built-in Python function to remove a specific character at a given index from a string. However, you can achieve this by using string slicing.

Here's an example code snippet that removes the character at index idx from the string s:

css
Copy code
s = s[:idx] + s[idx+1:]
This code creates a new string by concatenating the slice of the original string up to the index idx with the slice of the original string after the index idx+1. This effectively removes the character at index idx.

Note that this code assumes that the index idx is valid for the string s, otherwise an IndexError will be raised.





```python
s = s[:idx] + s[idx+1:]
```

Word prefix reversal
1 PT

The function reverse_prefix gets a string s and a single letter letter. The function should find the first occurrence of letter in s and reverse only the letters from the beginning of the string up to and including the target letter. If the letter does not exist in the string, do nothing.

For example, if you get the string "hackreactor" with the letter "c", then it would find the first c in "hackreactor", take all of the letters from the beginning of the string up to and including that "c" ("hackreactor"), and reverse them ("cahkreactor"). The answer would be the string "cahkreactor".

s	letter	Output
abcdefd	d	dcbaefd
xyxzxe	z	zxyxxe
abcd	z	abcd
abcd	a	abcd

In [2]:
def reverse_prefix(s, letter):
    for i in range(len(s)):
        if s[i] == letter:
            return s[:i+1][::-1] + s[i+1:]
    return s

### it cannot be `s[:i+1:-1] ` and has to be `s[:i+1][::-1] `

In [3]:
def is_double_reversible(num):
    if num < 0:
        return False
    rev_num = 0
    temp_num = num
    while temp_num > 0:
        rev_num = rev_num * 10 + temp_num % 10
        temp_num //= 10
    while rev_num > 0:
        if num % 10 != rev_num % 10:
            return False
        num //= 10
        rev_num //= 10
    return True

is_double_reversible(526) => True
is_double_reversible(1800) => False
is_double_reversible(0) => True
is_double_reversible(770) => False
is_double_reversible(-123) => False
is_double_reversible(12321) => True
is_double_reversible(12345) => False
is_double_reversible(987654321) => False
is_double_reversible(123456789) => False
is_double_reversible(1221) => True
is_double_reversible(123321) => True

In [4]:
is_double_reversible(526) 

False

In [5]:
is_double_reversible(525) 

True

In [6]:
def bubble_sort(values):
    n = len(values)
    
    # Traverse through all array elements
    for i in range(n):
        # Last i elements are already sorted
        for j in range(0, n-i-1):
            # Swap if the element found is greater than the next element
            if values[j] > values[j+1]:
                values[j], values[j+1] = values[j+1], values[j]
                
    return values

The provided implementation of Bubble Sort is already efficient for small-sized lists, but we can make some improvements to optimize it further. Here are a few optimizations that we can apply:

Early termination: If the inner loop goes through the entire list without any swap, then the list is already sorted. In such a case, we can terminate the sorting early to save unnecessary iterations.

Limit the range of the inner loop: After each outer loop iteration, the largest unsorted element gets placed at the end of the list. Therefore, we can reduce the range of the inner loop for each subsequent iteration.

Track if the list is already sorted: We can keep track of whether any swap has occurred in the inner loop. If no swaps occurred, then we can terminate the sorting early, as the list is already sorted.

Here's an optimized implementation of Bubble Sort that applies these optimizations:

python
Copy code
def bubble_sort(values):
    n = len(values)
    swapped = True
   ```python 
    # Traverse through all array elements
    for i in range(n):
        # Last i elements are already sorted
        swapped = False
        for j in range(0, n-i-1):
            # Swap if the element found is greater than the next element
            if values[j] > values[j+1]:
                values[j], values[j+1] = values[j+1], values[j]
                swapped = True
        
        # If no two elements were swapped by inner loop, then break
        if not swapped:
            break
                
    return values
   ```

In [9]:
s = 'abcdefg'
s[:3][::-1]

'cba'

In [10]:
s[:3:-1]

'gfe'

# Three Sum

In [1]:
def threeNumberSum(array, targetSum):
    res = []
    two_num_sum = []
    for n in array:
        two_num_sum.append(targetSum - n)
    i = 0
    while i < len(array):
        j = i+1
        while j < len(array):
            if array[i] + array[j] in two_num_sum:
                idx = two_num_sum.index(array[i] + array[j])
                if idx not in (i,j):
                    lst = sorted([array[idx], array[i], array[j]])
                    if lst not in res:
                        res.append(lst)
            j+=1
        i+=1
    return sorted(res)

# Quicksort

In [ ]:
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        self.quicksort(nums, 0, len(nums) - 1)
        return nums
    
    def quicksort(self, nums, lower, upper):
        if lower < upper:
            pivot = self.partition(nums, lower, upper)
            self.quicksort(nums, lower, pivot - 1)
            self.quicksort(nums, pivot + 1, upper)
        else:
            return
        
    def partition(self, nums, lower, upper):
        
        pivot = nums[upper]
        
        i = lower
        
        for j in range(lower, upper):
            if nums[j] < pivot:
                nums[i], nums[j] = nums[j], nums[i]
                i += 1
                
        nums[i], nums[upper] = nums[upper], nums[i]
        
        return i

# Insert SQL Function

In [ ]:
CREATE OR REPLACE FUNCTION generate_order_meals()
RETURNS void AS
$$
DECLARE
    order_id integer;
    meal_id integer;
    num_records integer;
BEGIN
    FOR order_id IN 1..50 LOOP
        num_records := floor(random() * 6) + 3; -- generate a random number of records between 3 to 8
        FOR i IN 1..num_records LOOP
            meal_id := floor(random() * 20) + 1; -- generate a random meal_id
            EXECUTE format('INSERT INTO order_meals (order_id, meal_id, quantity) VALUES (%s, %s, %s)',
                        order_id, meal_id, floor(random() * 5) + 1); -- generate a random quantity between 1 to 5
        END LOOP;
    END LOOP;
END;
$$
LANGUAGE plpgsql;

SELECT generate_order_meals();
                    

# Cocktail Sort

In [1]:
def cocktailSort(a):
    n = len(a)
    swapped = True
    start = 0
    end = n-1
    while (swapped==True):
 
        # reset the swapped flag on entering the loop,
        # because it might be true from a previous
        # iteration.
        swapped = False
 
        # loop from left to right same as the bubble
        # sort
        for i in range (start, end):
            if (a[i] > a[i+1]) :
                a[i], a[i+1]= a[i+1], a[i]
                swapped=True
 
        # if nothing moved, then array is sorted.
        if (swapped==False):
            break
 
        # otherwise, reset the swapped flag so that it
        # can be used in the next stage
        swapped = False
 
        # move the end point back by one, because
        # item at the end is in its rightful spot
        end = end-1
 
        # from right to left, doing the same
        # comparison as in the previous stage
        for i in range(end-1, start-1,-1):
            if (a[i] > a[i+1]):
                a[i], a[i+1] = a[i+1], a[i]
                swapped = True
 
        # increase the starting point, because
        # the last stage would have moved the next
        # smallest number to its rightful spot.
        start = start+1
 
# Driver code to test above
a = [5, 1, 4, 2, 8, 0, 2]
cocktailSort(a)
print("Sorted array is:")
for i in range(len(a)):
    print ("%d" %a[i]),

Sorted array is:
0
1
2
2
4
5
8


# First Unique Charactor

In [1]:
# Given a string, find the first non-repeating character in it and return its index. 
# If it doesn't exist, return -1. # Note: all the input strings are already lowercase.

#Approach 1
def solution(s):
    frequency = {}
    for i in s:
        if i not in frequency:
            frequency[i] = 1
        else:
            frequency[i] +=1
    for i in range(len(s)):
        if frequency[s[i]] == 1:
            return i
    return -1

print(solution('alphabet'))
print(solution('barbados'))
print(solution('crunchy'))

print('###')

#Approach 2
import collections

def solution(s):
    # build hash map : character and how often it appears
    count = collections.Counter(s) # <-- gives back a dictionary with words occurrence count 
                                         #Counter({'l': 1, 'e': 3, 't': 1, 'c': 1, 'o': 1, 'd': 1})
    # find the index
    for idx, ch in enumerate(s):
        if count[ch] == 1:
            return idx     
    return -1

print(solution('alphabet'))
print(solution('barbados'))
print(solution('crunchy'))

1
2
1
###
1
2
1


# Multiply Strings

In [2]:
def multiply(num1, num2):
    product = [0] * (len(num1) + len(num2))
    pos = len(product)-1
    
    for n1 in reversed(num1):
        tempPos = pos
        for n2 in reversed(num2):
            product[tempPos] += int(n1) * int(n2)
            product[tempPos-1] += product[tempPos]/10
            product[tempPos] %= 10
            tempPos -= 1
        pos -= 1
        
    pt = 0
    while pt < len(product)-1 and product[pt] == 0:
        pt += 1

    return ''.join(map(str, product[pt:]))